Refactor custom filters

smalluban opened the pull request:
WIP

smalluban commented on extension-manifest-v3/src/background/custom-filters.js:
```diff
@@ -31,51 +123,88 @@ async function updateDNRRules(dnrRules) {
+  };
+}
+async function update(text, { trustedScriptlets }) {
+  const parseResult = parseInput(text, { trustedScriptlets });
```
I tried to move the logic without changes, but I am not sure if this is still required - as for the engine, the input of the filters is parsed again, this time by the `parseFilters` function from the adblocker.

I left it here for now, as the internal parser uses the `trustedScriptlets` flag - but there was a bug before, which did not pass "fixed" scriptlets to the engine, but it generated them again.

philipp-classen commented on extension-manifest-v3/src/store/options.js:
```diff
@@ -303,19 +318,36 @@ async function migrateFromV8() {
   }
 }
 
+function isEqual(a, b) {
```
Note that it is not symmetric:
```
> isEqual({}, { foo: 'bar' })
true
> isEqual({ foo: 'bar' }, {})
false
```

smalluban replied:
Good catch, but it is safe here - as the structure of both compared objects is always identical - they are both two versions of options. I think I'll rename the function to `isEqualOptions` to avoid confusion.

philipp-classen commented on extension-manifest-v3/src/background/custom-filters.js:
```diff
@@ -8,10 +8,102 @@
+  return `${front}#+js(${[parsedScript.name, ...args].join(', ')})`;
+}
+
+function parseInput(text = '', { trustedScriptlets }) {
```
IMHO, the old name "parseFilters" was more descriptive than "parseInput".

smalluban replied:
The function is not required anymore - the only part with fixing scriptlets will be moved, after discussion with @chrmod

philipp-classen commented on extension-manifest-v3/src/utils/dnr-converter.js:
```diff
@@ -18,12 +18,16 @@ export function createDocumentConverter() {
-      if (__PLATFORM__ === 'safari') {
-        event.data.rules = event.data.rules.map(getCompatRule).filter(Boolean);
-      }
+      requests.get(requestId)({
```
Here I struggled first to understand what the effect is (same with the original version). Took me a bit to realize that "requests" is a mapping to callbacks, which should be triggered when the results of a conversion are available.

Maybe it would help the reader to name it to something like "requestResolvers"?

smalluban replied:
I'll look into it, but this part was at least touched by me in this PR ;)

smalluban commented:
@philipp-classen Updated and fixed what I could change. The `parseInput` must stay, but I renamed it to `normalizeFilters`, as it better describes what actually the function does - it does not produce filters, just split the input into buckets, and fixes scriptlets if needed.

philipp-classen commented on extension-manifest-v3/src/background/custom-filters.js:
```diff
@@ -6,7 +6,7 @@
  * This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0
+w * file, You can obtain one at http://mozilla.org/MPL/2.0
```
A "w" slipped in by accident.

smalluban commented:
@GRadziejewski Testing highlights:
* The feature logic has not changed
* Added on/off toggle for the whole feature, so we can disable custom filters without removing the input
* Fixed issue with updated adblocker engine - to proof the bug update 3.18 -> 4.0, to proof the fix 3.18 -> 4.0 -> this PR

GRadziejewski commented:
Test results: https://github.com/ghostery/ghostery-extension/issues/1822#issuecomment-2288995414

Issue: https://github.com/ghostery/ghostery-extension/issues/1823

smalluban commented:
After testing the feature, I added a small change to not confuse users - the "Update" (actually "Save" now - it's more what it is, and it's already translated phrase) is disabled only when filters update.
Also, it fixes the state when the global pause is turned on.

GRadziejewski commented:
Test results: https://github.com/ghostery/ghostery-extension/issues/1822

LGTM.
